posts/inside-rust/2024-05-01-enabling-rust-lld-on-linux.md

```diff
@@ -0,0 +1,83 @@
+installed on the user's machine.
+
+Since improvements to linking times are substantial, it would be a good default to use in the most
+popular targets. This has been discussed for a long time, for example in issues
+[#39915](https://github.com/rust-lang/rust/issues/39915) and
+[#71515](https://github.com/rust-lang/rust/issues/71515), and rustc already offers nightly flags to
+use `rust-lld`.
+
+By now, we believe we've done all the internal testing that we could, on CI, crater, our
+benchmarking infrastructure, and would like to expand testing and gather real-world feedback and
+use-cases. Therefore, we will enable `rust-lld` to be the linker used by default on
+`x86_64-unknown-linux-gnu` for nightly builds.
+
+#### Benefits
+
+While this also enables the compiler to use more linker features in the future, the most immediate
+benefit is much improved linking times.
+
+Here are more details from the ripgrep example mentioned above: linking is reduced 7x, resulting in
+a 40% reduction in end-to-end compilation times.
+
+![Before/after comparison of a `ripgrep` debug build](../../../../images/inside-rust/2024-05-01-enabling-rust-lld-on-linux/ripgrep-comparison.png)
+
+Most binaries should see some improvements here, but it's especially significant with e.g. bigger
+binaries, or when involving debuginfo. These usually see bottlenecks in the linker.
+
+Here's [a
+link](https://perf.rust-lang.org/compare.html?start=b3e117044c7f707293edc040edb93e7ec5f7040a&end=baed03c51a68376c1789cc373581eea0daf89967&stat=instructions%3Au&tab=compile)
+to the complete results from our benchmarks.
+
+If testing goes well, we can then stabilize using this faster linker by default for
+`x86_64-unknown-linux-gnu` users, before maybe looking at other targets.
+
+#### Possible drawbacks
+
+From our prior testing, we don't really expect issues to happen in practice. it is a drop-in
+replacement for the vast majority of cases, but `lld` is not _bug-for-bug_ compatible with GNU ld.
+
+In any case, using `rust-lld` can be disabled if any problem occurs: use the `-Z
+linker-features=-lld` flag to revert to using the system's default linker.
+
+Some crates somehow relying on these differences could need additional link args. For example, we
+saw <20 crates in the crater run failing to link because of a different default about [encapsulation
+symbols](https://lld.llvm.org/ELF/start-stop-gc): these could require
+`-Clink-arg=-Wl,-z,nostart-stop-gc` to match the legacy GNU ld behavior.
+
+Some of the big gains in performance come from parallelism, which could be undesirable in
+resource-constrained environments.
+
+#### Summary
+
+rustc will use `rust-lld` on `x86_64-unknown-linux-gnu` nightlies, for much improved linking times.
+Let us know if you encounter problems, by opening an issue on github. If that happens, you can
+revert to the default linker with the `-Z linker-features=-lld` flag.
```

Comment:
Maybe we can say something about how to do that with Cargo? e.g., RUSTFLAGS is the right thing, but maybe we should mention `.cargo/config.toml` or similar so they don't need to set it globally?

Reply:
I've pushed afbb1fd88dd7b670181b21d390cab2b2bda2fe2d to address this.

On the opposite side of this coin, we could also mention `readelf -p .comment` to check if lld is indeed used to link a given binary.